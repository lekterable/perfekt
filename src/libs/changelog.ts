import {
  generateChangelog,
  readFile,
  writeFile,
  fileExists,
  printOutput,
  createReleasedFilter
} from '~utils'
import { defaults, Optionable } from './options'
import { Config } from '~types'
import { ChangelogOptions, GroupedCommits } from '~types'

class Changelog extends Optionable<ChangelogOptions> {
  #config

  constructor(config: Config) {
    super(defaults.changelogOptions)
    this.#config = config
  }

  get hasChangelog() {
    return fileExists('CHANGELOG.md')
  }

  async getChangelog(): Promise<string> {
    return (await readFile('CHANGELOG.md')) as string
  }

  async getHistory(version: string) {
    const isFromOverwritten = Boolean(this.options.root || this.options.from)

    if (!this.hasChangelog || isFromOverwritten) return

    const changelog = await this.getChangelog()

    if (!changelog) {
      throw new Error("CHANGELOG doesn't contain any history.")
    }

    const filterReleased = createReleasedFilter(version, this.#config)

    return changelog.split('\n').filter(filterReleased).join('\n')
  }

  generate(commits: GroupedCommits, version: string | undefined) {
    return generateChangelog(commits, version, this.#config)
  }

  save(changelog: string) {
    return writeFile('CHANGELOG.md', changelog)
  }

  print(changelog: string) {
    printOutput(changelog)
  }

  async finish(
    commits: GroupedCommits,
    version: string | undefined,
    previousVersion: string | undefined
  ) {
    let changelog = this.generate(commits, version)

    if (previousVersion) {
      const history = await this.getHistory(previousVersion)

      if (history) changelog = changelog + '\n' + history
    }

    if (this.options.write) return this.save(changelog)

    return this.print(changelog)
  }
}

export default Changelog
