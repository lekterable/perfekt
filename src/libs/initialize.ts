import inquirer from 'inquirer'
import { writeFile } from '~utils'

export const questions = [
  {
    type: 'list',
    name: 'configFormat',
    message: 'Which format do you want to use for your your config?',
    choices: [
      '.perfektrc',
      'perfekt.config.js',
      'perfekt.yaml',
      'perfekt.json'
    ],
    default: '.perfektrc'
  }
]

const initialize = async () => {
  const { configFormat } = await inquirer.prompt(questions)

  writeFile(configFormat, '')
}

export default initialize
