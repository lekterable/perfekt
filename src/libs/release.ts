import { defaults, Optionable } from './options'
import Git from './git'
import NPM from './npm'
import { ReleaseOptions } from '~types'

class Release extends Optionable<ReleaseOptions> {
  #git
  #npm

  constructor() {
    super(defaults.releaseOptions)
    this.#npm = new NPM()
    this.#git = new Git()
  }

  finish(version: string) {
    this.#npm.updateVersion(version)
    this.#git.release(version)
  }
}

export default Release
