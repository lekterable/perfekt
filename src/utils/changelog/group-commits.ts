import { Commit } from '~libs'
import isObjectEmpty from '../object/is-empty'
import { Config, GroupedCommits } from '~types'

const groupCommits = (commits: Commit[], config: Config): GroupedCommits =>
  commits.reduce(
    (grouped, commit) => {
      const group = grouped[grouped.length - 1]
      const rest = grouped.slice(0, -1)

      if (commit.isConventional) {
        const normalizedScope = commit.scope?.toLowerCase()

        if (normalizedScope && config.ignoredScopes.includes(normalizedScope))
          return [...grouped]

        if (normalizedScope === 'release') {
          const isLatest = isObjectEmpty(group)
          const release = { release: commit }

          return isLatest ? [release] : [...grouped, release]
        }

        if (commit.isBreaking) {
          // @ts-expect-error -- Make type-safe later
          const existing = group.breaking ? group.breaking : []
          return [...rest, { ...group, breaking: [...existing, commit] }]
        }
      }

      const matchingGroup = config.groups.find(({ types }) =>
        // @ts-expect-error -- Make type-safe later
        types.includes(commit.type)
      )

      if (!matchingGroup) {
        // @ts-expect-error -- Make type-safe later
        const existing = group.misc ? group.misc : []
        return [...rest, { ...group, misc: [...existing, commit] }]
      }

      const key = matchingGroup.types[0]
      // @ts-expect-error -- Make type-safe later
      const existing = group[key] ? group[key] : []
      return [...rest, { ...group, [key]: [...existing, commit] }]
    },
    [{}]
  )

export default groupCommits
