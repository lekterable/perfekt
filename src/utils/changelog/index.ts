export { default as createReleasedFilter } from './create-released-filter'
export { default as generateChangelog } from './generate-changelog'
export { default as generateLine } from './generate-line'
export { default as groupCommits } from './group-commits'
