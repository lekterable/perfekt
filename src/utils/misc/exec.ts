import process from 'child_process'

const exec = (command: string): Buffer | string | undefined =>
  process.execSync(command)

export default exec
