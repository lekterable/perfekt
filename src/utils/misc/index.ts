export { default as exec } from './exec'
export { default as readFile } from './read-file'
export { default as writeFile } from './write-file'
export { default as printOutput } from './print-output'
export { default as fileExists } from './file-exists'
