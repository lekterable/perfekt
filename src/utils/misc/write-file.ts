import fs from 'fs'

const writeFile = (fileName: string, content: string) =>
  new Promise<void>((resolve, reject) =>
    fs.writeFile(fileName, content, 'utf8', err => {
      if (err) return reject(err)

      return resolve()
    })
  )

export default writeFile
