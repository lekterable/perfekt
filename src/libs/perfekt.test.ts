import fs from 'fs'
import inquirer from 'inquirer'
import Git from './git'
import Config from '~config'
import Perfekt from './perfekt'
import * as groupCommits from '~utils/changelog/group-commits'
import * as determineVersion from '~utils/release/determine-version'
import * as initialize from './initialize'
import * as writeFile from '~utils/misc/write-file'
import Release from './release'
import Changelog from './changelog'
import Options, { defaults } from './options'
import exec from '~utils/misc/exec'
import Commit from './commit'

jest.mock('fs')
jest.mock('inquirer')
jest.mock('~utils/misc/exec')

const fsMock = jest.mocked(fs)
const inquirerMock = jest.mocked(inquirer)
const execMock = jest.mocked(exec)

describe('Perfekt', () => {
  const { config } = new Config()
  const perfekt = new Perfekt(config)

  const mockLog =
    'bffc2f9e8da1c7ac133689bc9cd14494f3be08e3 refactor: extract line generating logic to function and promisify exec\naa805ce71ee103965ce3db46d4f6ed2658efd08d feat: add option to write to local CHANGELOG file\nb2f5901922505efbfb6dd684252e8df0cdffeeb2 fix: support other conventions\nbffc2f9e8da1c7ac133689bc9cd14494f3be08e3 fix: a bug\n4e02179cae1234d7083036024080a3f25fcb52c2 feat: add execute release feature\nb2f5901922505efbfb6dd684252e8df0cdffeeb2 tests: add core tests\n2ea04355c1e81c5088eeabc6e242fb1ade978524 chore: update dependencies'

  const mockGrouped = [
    {
      feat: [
        new Commit(
          'aa805ce71ee103965ce3db46d4f6ed2658efd08d feat: add option to write to local CHANGELOG file'
        ),
        new Commit(
          '4e02179cae1234d7083036024080a3f25fcb52c2 feat: add execute release feature'
        )
      ],
      fix: [
        new Commit(
          'nb2f5901922505efbfb6dd684252e8df0cdffeeb2 fix: support other conventions'
        ),
        new Commit('bffc2f9e8da1c7ac133689bc9cd14494f3be08e3 fix: a bug')
      ],
      misc: [
        new Commit(
          'bffc2f9e8da1c7ac133689bc9cd14494f3be08e3 refactor: extract line generating logic to function and promisify exec'
        ),
        new Commit(
          'b2f5901922505efbfb6dd684252e8df0cdffeeb2 tests: add core tests'
        ),
        new Commit(
          '2ea04355c1e81c5088eeabc6e242fb1ade978524 chore: update dependencies'
        )
      ]
    }
  ]

  const mockCommits = mockLog.split('\n').map(commit => new Commit(commit))

  describe('init', () => {
    let writeFileSpy: jest.SpiedFunction<typeof writeFile.default>
    let initializeSpy: jest.SpiedFunction<typeof initialize.default>

    beforeEach(() => {
      writeFileSpy = jest.spyOn(writeFile, 'default')
      initializeSpy = jest.spyOn(initialize, 'default')
    })

    it('should initialize configuration', async () => {
      const mockConfigFormat = '.perfektrc'

      inquirerMock.prompt.mockResolvedValue({ configFormat: mockConfigFormat })

      await perfekt.init()

      expect(initializeSpy).toBeCalledTimes(1)
      expect(writeFileSpy).toBeCalledTimes(1)
      expect(writeFileSpy).toBeCalledWith(mockConfigFormat, '')
    })
  })

  describe('release', () => {
    const releaseOptions = new Options(defaults.releaseOptions)

    let changelogSpy: jest.SpiedFunction<typeof Perfekt.prototype.changelog>
    let finishSpy: jest.SpiedFunction<typeof Release.prototype.finish>
    let determineVersionSpy: jest.SpiedFunction<typeof determineVersion.default>

    beforeEach(() => {})

    beforeEach(() => {
      releaseOptions.restore()

      changelogSpy = jest.spyOn(Perfekt.prototype, 'changelog')
      finishSpy = jest.spyOn(Release.prototype, 'finish')
      determineVersionSpy = jest.spyOn(determineVersion, 'default')
    })

    it('should execute a release', async () => {
      const mockVersion = '2.2.2'
      const options = releaseOptions.options

      execMock.mockReturnValueOnce(mockVersion)
      execMock.mockReturnValueOnce(mockLog)
      // @ts-expect-error -- Make type-safe later
      fsMock.writeFile.mockImplementationOnce((_, __, ___, cb) => cb(undefined))

      await perfekt.release(mockVersion, options)

      expect(determineVersionSpy).toBeCalledTimes(1)
      expect(determineVersionSpy).toBeCalledWith(mockVersion, config)
      expect(changelogSpy).toBeCalledTimes(1)
      expect(changelogSpy).toBeCalledWith(mockVersion, {
        from: undefined,
        root: false,
        write: true
      })
      expect(finishSpy).toBeCalledTimes(1)
      expect(finishSpy).toBeCalledWith(mockVersion)
    })

    it('should execute a release with `--from` option', async () => {
      const mockVersion = '2.2.2'
      const mockSHA = '4e02179c'
      const options = releaseOptions.merge({ from: mockSHA }).options

      execMock.mockReturnValueOnce(mockVersion)
      execMock.mockReturnValueOnce(mockLog)
      // @ts-expect-error -- Make type-safe later
      fsMock.writeFile.mockImplementationOnce((_, __, ___, cb) => cb(undefined))

      await perfekt.release(mockVersion, options)

      expect(determineVersionSpy).toBeCalledTimes(1)
      expect(determineVersionSpy).toBeCalledWith(mockVersion, config)
      expect(changelogSpy).toBeCalledTimes(1)
      expect(changelogSpy).toBeCalledWith(mockVersion, {
        from: mockSHA,
        root: false,
        write: true
      })
      expect(finishSpy).toBeCalledTimes(1)
      expect(finishSpy).toBeCalledWith(mockVersion)
    })
  })

  describe('changelog', () => {
    const changelogOptions = new Options(defaults.changelogOptions)

    let printSpy: jest.SpiedFunction<typeof Changelog.prototype.print>
    // @ts-expect-error -- Make type-safe later
    let latestTagSpy: jest.SpiedFunction<typeof Git.prototype.latestTag>
    let finishSpy: jest.SpiedFunction<typeof Changelog.prototype.finish>
    let getCommitsSpy: jest.SpiedFunction<typeof Git.prototype.getCommits>
    let groupCommitsSpy: jest.SpiedFunction<typeof groupCommits.default>

    beforeEach(() => {
      changelogOptions.restore()

      printSpy = jest.spyOn(Changelog.prototype, 'print')
      // @ts-expect-error -- Make type-safe later
      latestTagSpy = jest.spyOn(Git.prototype, 'latestTag', 'get')
      finishSpy = jest.spyOn(Changelog.prototype, 'finish')
      getCommitsSpy = jest.spyOn(Git.prototype, 'getCommits')
      groupCommitsSpy = jest.spyOn(groupCommits, 'default')
    })

    it('should generate changelog', async () => {
      const mockVersion = '2.2.2'
      const options = changelogOptions.options

      execMock.mockReturnValueOnce(mockVersion)
      execMock.mockReturnValueOnce(mockLog)
      printSpy.mockImplementation()

      await perfekt.changelog(mockVersion, options)

      expect(latestTagSpy).toBeCalledTimes(1)
      expect(getCommitsSpy).toBeCalledTimes(1)
      expect(getCommitsSpy).toBeCalledWith(mockVersion)
      expect(groupCommitsSpy).toBeCalledTimes(1)
      expect(groupCommitsSpy).toBeCalledWith(mockCommits, config)
      expect(finishSpy).toBeCalledTimes(1)
      expect(finishSpy).toBeCalledWith(mockGrouped, mockVersion, mockVersion)
      expect(printSpy.mock.calls[0][0]).toMatchSnapshot()
    })

    it('should generate changelog with `--root` option', async () => {
      const mockVersion = '2.2.2'
      const options = changelogOptions.merge({ root: true }).options

      execMock.mockReturnValueOnce(mockVersion)
      execMock.mockReturnValueOnce(mockLog)
      printSpy.mockImplementation()

      await perfekt.changelog(mockVersion, options)

      expect(latestTagSpy).toBeCalledTimes(1)
      expect(getCommitsSpy).toBeCalledTimes(1)
      expect(getCommitsSpy).toBeCalledWith(undefined)
      expect(groupCommitsSpy).toBeCalledTimes(1)
      expect(groupCommitsSpy).toBeCalledWith(mockCommits, config)
      expect(finishSpy).toBeCalledTimes(1)
      expect(finishSpy).toBeCalledWith(mockGrouped, mockVersion, mockVersion)
      expect(printSpy.mock.calls[0][0]).toMatchSnapshot()
    })

    it('should generate changelog with `--from` option', async () => {
      const mockVersion = '2.2.2'
      const mockSHA = '4e02179c'
      const options = changelogOptions.merge({ from: mockSHA }).options

      execMock.mockReturnValueOnce(mockVersion)
      execMock.mockReturnValueOnce(mockLog)
      printSpy.mockImplementation()

      await perfekt.changelog(mockVersion, options)

      expect(latestTagSpy).toBeCalledTimes(1)
      expect(getCommitsSpy).toBeCalledTimes(1)
      expect(getCommitsSpy).toBeCalledWith(mockSHA)
      expect(groupCommitsSpy).toBeCalledTimes(1)
      expect(groupCommitsSpy).toBeCalledWith(mockCommits, config)
      expect(finishSpy).toBeCalledTimes(1)
      expect(finishSpy).toBeCalledWith(mockGrouped, mockVersion, mockVersion)
      expect(printSpy.mock.calls[0][0]).toMatchSnapshot()
    })
  })
})
