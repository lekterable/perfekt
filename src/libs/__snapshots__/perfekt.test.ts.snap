// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Perfekt changelog should generate changelog 1`] = `
"# 2.2.2

## Features

- add option to write to local CHANGELOG file aa805ce7
- add execute release feature 4e02179c

## Fixes

- support other conventions b2f59019
- a bug bffc2f9e

## Misc

- extract line generating logic to function and promisify exec bffc2f9e
- add core tests b2f59019
- update dependencies 2ea04355
"
`;

exports[`Perfekt changelog should generate changelog with \`--from\` option 1`] = `
"# 2.2.2

## Features

- add option to write to local CHANGELOG file aa805ce7
- add execute release feature 4e02179c

## Fixes

- support other conventions b2f59019
- a bug bffc2f9e

## Misc

- extract line generating logic to function and promisify exec bffc2f9e
- add core tests b2f59019
- update dependencies 2ea04355
"
`;

exports[`Perfekt changelog should generate changelog with \`--root\` option 1`] = `
"# 2.2.2

## Features

- add option to write to local CHANGELOG file aa805ce7
- add execute release feature 4e02179c

## Fixes

- support other conventions b2f59019
- a bug bffc2f9e

## Misc

- extract line generating logic to function and promisify exec bffc2f9e
- add core tests b2f59019
- update dependencies 2ea04355
"
`;
